### PR TITLE
[backend-common, core-app-api] Add HostDiscovery to read plugin baseUrls from config

### PR DESCRIPTION
**Pull request by UsainBloot**
## Context
Currently, to be able to point to different backend endpoints via the `discoveryApi`, the implementation has to be re-written. This PR adds the ability to declare plugin discovery URLs in config rather than code within `@backstage/backend-common`. Also enables different endpoints to be used for environments based on relevant `app-config.yaml`.

**Proposed Config**
```yaml
discovery:
  endpoints:
    - target: https://internal.example.com/internal-catalog
      plugins: [catalog]
    - target: https://internal.example.com/secure/api/{{pluginId}}
      plugins: [auth, permissions]
    - target:
        internal: https://internal.example.com/search
        external: https://example.com/search
      plugins: [search]
```

Aligned `{{pluginId}}` replacement with pattern in [`urlPatternDiscovery.compile`](https://backstage.io/docs/reference/core-app-api.urlpatterndiscovery.compile)

### `backend-common`
- Added `HostDiscovery` to supersede `SingleHostDiscovery`
- **Deprecated** `SingleHostDiscovery` (as its name meant the functionality should not belong here)
- HostDiscovery will check the backend config for plugin specific baseUrls before using single host
### `core-app-api`
- Added `HostDiscovery` for a config driven discovery implementation

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts
Comment (Rugvip) [inline packages/backend-common/src/discovery/SingleHostDiscovery.ts:32]: Let's just rename this one instead of duplicating the implementation. Then re-export with `export const SingleHostDiscovery = HostDiscovery`, and mark that as deprecated instead

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (Rugvip) [inline packages/backend-common/src/discovery/HostDiscovery.ts:72]: Let's add config schema for this in `packages/backend-common/config.d.ts`

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (Rugvip) [inline packages/backend-common/src/discovery/HostDiscovery.ts:84]: I'm feeling that breaking this down by endpoints or so might be the way to go, rather than per plugin. Perhaps something like this?

```
discovery:
  endpoints:
    - target: https://internal.example.com/internal-catalog
      plugins: [catalog]
    - target: https://internal.example.com/secure/api/{pluginId}
      plugins: [auth, permissions]
    - target:
        internal: https://internal.example.com/search
        external: https://example.com/search
      plugins: [search]
```

All examples and throwing out some extra ideas, I don't think we should implement all of those capabilities right now but it's a couple of things for the future. What do you think?

**Review by Rugvip — COMMENTED** [review]
🎉

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/SingleHostDiscovery.ts]: Addressed in b58a2b7

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (Rugvip) [inline packages/backend-common/src/discovery/HostDiscovery.ts:39]: Noting that I don't think we want this under the backend config key, but rather a top-level `discovery` key, given that we'll want to be consuming this from the frontend as well

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (Rugvip) [inline packages/backend-common/src/discovery/HostDiscovery.ts:2]: Keep this as is

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in 54ea03b

File: packages/backend-common/src/discovery/index.ts
Comment (UsainBloot) [inline packages/backend-common/src/discovery/index.ts:29]: Currently causes the following issue
```
 ReferenceError: Cannot access 'HostDiscovery' before initialization
```

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in 960c1cf

Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: I've slightly changed the pattern to be `{{pluginId}}` or `{{ pluginId }}` to align with [`urlPatternDiscovery.compile`](https://backstage.io/docs/reference/core-app-api.urlpatterndiscovery.compile)

File: packages/backend-common/config.d.ts
Comment (Rugvip) [inline packages/backend-common/config.d.ts:225]: @backstage/maintainers what do you think about this format?

**Review by Rugvip — APPROVED** [review]
Nice, thank you! 😁 

Let's hold off a bit on shipping, wanna make sure we think about the format a lil bit more

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (freben) [inline packages/backend-common/src/discovery/HostDiscovery.ts:104]: just undefined - let's avoid null

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (freben) [inline packages/backend-common/src/discovery/HostDiscovery.ts:108]:
```suggestion
      return target.replace(/\{\{\s*pluginId\s*\}\}/g, pluginId);
```

I know it's a silly thing but hey, why not, if you want to

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (freben) [inline packages/backend-common/src/discovery/HostDiscovery.ts:111]:
```suggestion
    return target[type].replace(/\{\{\s*pluginId\s*\}\}/g, pluginId);
```

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (freben) [inline packages/backend-common/src/discovery/HostDiscovery.ts:121]: I just realized, we should url encode the plugin id just to be well behaved

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (freben) [inline packages/backend-common/src/discovery/HostDiscovery.ts:131]: encode here too

Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/backend-common/src/discovery/HostDiscovery.ts
Comment (UsainBloot) [reply packages/backend-common/src/discovery/HostDiscovery.ts]: Addressed in [1364df6](https://github.com/backstage/backstage/pull/17152/commits/1364df62a19fda7c2964c0476fa6d21ad34e7ba0)

File: packages/core-app-api/src/apis/implementations/DiscoveryApi/HostDiscovery.ts
Comment (UsainBloot) [inline packages/core-app-api/src/apis/implementations/DiscoveryApi/HostDiscovery.ts:81]: Just wanted to call this out.

Is using the `external` target the correct approach if there is both an `internal` and `external` target?

File: packages/core-app-api/src/apis/implementations/DiscoveryApi/HostDiscovery.ts
Comment (Rugvip) [reply packages/core-app-api/src/apis/implementations/DiscoveryApi/HostDiscovery.ts]: Ended up with a bit of a refactor to change this and some other bits in https://github.com/backstage/backstage/pull/17152/commits/51329677e184472f045d184a31c16244cf425e15. Main thing I wanted to keep from the existing implementation was the error handling, but also the pre-parsing of the patterns for a tiny bit of perf. I changed the implementation to use the existing `UrlPatternDiscovery` directly instead. Also renamed to `FrontendHostDiscovery` as I feel having symbols exported from both frontend and backend packages is a recipe for massive confusion 😅

**Pull request merged**
